src/x-server-local.c

```diff
@@ -391,6 +394,44 @@ got_signal_cb (Process *process, int signum, XServerLocal *server)
     }
 }
 
+static gboolean
+poll_for_socket_cb (XServerLocal *server)
+{
+    XServerLocalPrivate *priv = x_server_local_get_instance_private (server);
+
+    /* Check is X11 socket file exists as an alternative startup test to SIGUSR1 */
+    GStatBuf statbuf;
+    g_autofree gchar *socketpath = g_strdup_printf ("/tmp/.X11-unix/X%d", priv->display_number);
+
+    if ( g_stat (socketpath, &statbuf) == 0 )
+    {
+        uid_t uid = priv->user ? user_get_uid(priv->user) : 0;
+
+        /* It has to be a valid socket file */
+        if (!(statbuf.st_mode & S_IFSOCK))
```

Comment:
The if statement is a little hard to follow here - I think it would be simpler if you just return from the function, i.e.

```
if (!(statbuf.st_mode & S_IFSOCK))
   return G_SOURCE_REMOVE;
```

Reply:
I agree it is a bit hard to read. I've changed it as you suggested.